### PR TITLE
style(ruff): fix linting issues from ruff update

### PR DESCRIPTION
**Pull request by lengau**
- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox`?

-----

### CODE REVIEW

**Review by upils — CHANGES_REQUESTED** [review]
This PR and https://github.com/canonical/craft-application/pull/575 solve the same issue in different ways. 
In the end I think this one looks better because using `.items()` avoid getting the `unprocessed_grammar` with the `key`, and avoid disabling `PLC0206` (I suspect we will forget to re-enable it later).

But I think this PR can be updated to bump the ruff version in `.pre-commit-config.yaml`.

**Review by tigarmo — APPROVED** [review]

File: craft_application/grammar.py [hunk 1]
```diff
@@ -79,7 +77,8 @@ def process_part(
             unprocessed_grammar = cast(list[dict[str, Any] | str], unprocessed_grammar)
         # grammar aware models can be a string
         elif isinstance(unprocessed_grammar, str):
-            unprocessed_grammar = [unprocessed_grammar]
+            # Ignore rule PLW2901 because we do need to overwrite a for loop variable.
+            unprocessed_grammar = [unprocessed_grammar]  # noqa: PLW2901
```
Comment (dariuszd21) [inline craft_application/grammar.py:81]: If you have to silent `ruff` anyway isn't it better to just silent `PLC0206` at the loop level?
Alternatively you could do an assignment in line 56, to leave logic intact.
I'm not a big fun of shadowing the variable inside the loop, as it makes it confusing to follow the changes (and it's easy to overlook that item was actually processed)
```python
for key, part_data in part_yaml_data.items():
        unprocessed_grammar = part_data
```

**Review by dariuszd21 — APPROVED** [review]
Thank you!
It seems that we've done the same as Paul already discovered :sweat_smile:

File: craft_application/grammar.py (hunk 1 above)
Comment (lengau) [reply craft_application/grammar.py]: Done!

**Review by upils — APPROVED** [review]

**Pull request merged**